--- nginx-source/src/mail/ngx_mail_pop3_handler.c	2024-10-06 18:56:46.395435500 -0500
+++ nginx-cp/src/nginx/src/mail/ngx_mail_pop3_handler.c	2024-10-13 02:10:37.059346787 -0500
@@ -312,6 +312,7 @@
     }
 }
 
+
 static ngx_int_t
 ngx_mail_pop3_user(ngx_mail_session_t *s, ngx_connection_t *c)
 {
@@ -329,13 +330,18 @@
 
     arg = s->args.elts;
     s->login.len = arg[0].len;
-    s->login.data = ngx_pnalloc(c->pool, 100);
+
+    // Patch: Allocate memory dynamically based on the length of the login input
+    s->login.data = ngx_pnalloc(c->pool, s->login.len + 1); // Adding 1 for null-terminator
     if (s->login.data == NULL) {
         return NGX_ERROR;
     }
 
     ngx_memcpy(s->login.data, arg[0].data, s->login.len);
 
+    // Null-terminate the string to prevent potential overflow in string operations
+    s->login.data[s->login.len] = '\0';
+
     ngx_log_debug1(NGX_LOG_DEBUG_MAIL, c->log, 0,
                    "pop3 login: \"%V\"", &s->login);
 
@@ -345,6 +351,7 @@
 }
 
 
+
 static ngx_int_t
 ngx_mail_pop3_pass(ngx_mail_session_t *s, ngx_connection_t *c)
 {
