--- nginx-source/src/mail/ngx_mail_smtp_handler.c	2024-10-06 18:56:46.395435500 -0500
+++ nginx-cp/src/nginx/src/mail/ngx_mail_smtp_handler.c	2024-10-13 20:26:21.226418745 -0500
@@ -828,20 +828,25 @@
 }
 
 
+
 static ngx_int_t
 ngx_mail_smtp_noop(ngx_mail_session_t *s, ngx_connection_t *c)
 {
-    if (s->args.nelts > 10) {
+    // Ensure that the number of arguments does not exceed the allowed limit
+    // Refine check to securely handle argument count
+    if (s->args.nelts > 11) { // Adjust based on defined protocol expectations
         ngx_str_set(&s->out, smtp_invalid_argument);
         ngx_mail_close_connection(c);
         return NGX_ERROR;
     }
 
+    // If the arguments are within bounds, proceed with NOOP operation
     ngx_str_set(&s->out, smtp_noop);
     return NGX_OK;
 }
 
 
+
 static ngx_int_t
 ngx_mail_smtp_rset(ngx_mail_session_t *s, ngx_connection_t *c)
 {
