--- nginx-source/src/http/ngx_http_core_module.c	2024-10-06 18:56:46.395435500 -0500
+++ nginx-cp/src/nginx/src/http/ngx_http_core_module.c	2024-10-13 21:01:54.444000196 -0500
@@ -1946,6 +1946,7 @@
 }
 
 
+
 ngx_int_t
 ngx_http_auth_basic_user(ngx_http_request_t *r)
 {
@@ -1985,7 +1986,8 @@
         return NGX_DECLINED;
     }
 
-    auth.len = NGX_HTTP_AUTH_MAX;
+    // Calculate the necessary length for the decoded output
+    auth.len = ngx_base64_decoded_length(encoded.len);
     auth.data = ngx_pnalloc(r->pool, auth.len + 1);
     if (auth.data == NULL) {
         return NGX_ERROR;
@@ -2018,6 +2020,7 @@
 }
 
 
+
 #if (NGX_HTTP_GZIP)
 
 ngx_int_t
